### ISSUE TITLE
Some event sending fail when POST request with non ASCII string

### ISSUE THREAD

**Issue by mkusaka** [getsentry/sentry-ruby#1475]
**Describe the bug**
When POST request through fetch with non ASCII string, error event failed to send to sentry server.

With `config.debug = true`, following log printed in console.

```
Event sending failed: "\xE3" from ASCII-8BIT to UTF-8
/Users/username/.rbenv/versions/2.7.3/lib/ruby/2.7.0/json/common.rb:224:in `generate'
/Users/username/.rbenv/versions/2.7.3/lib/ruby/2.7.0/json/common.rb:224:in `generate'
/Users/username/.rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/sentry-ruby-core-4.5.1/lib/sentry/transport.rb:102:in `encode'
/Users/username/.rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/sentry-ruby-core-4.5.1/lib/sentry/transport.rb:42:in `send_event'
/Users/username/.rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/sentry-ruby-core-4.5.1/lib/sentry/client.rb:86:in `send_event'
/Users/username/.rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/sentry-ruby-core-4.5.1/lib/sentry/client.rb:110:in `block in dispatch_background_event'
/Users/username/.rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/sentry-rails-4.5.1/lib/sentry/rails/background_worker.rb:7:in `block (2 levels) in perform'
...
```

**To Reproduce**
see: https://github.com/mkusaka/sentry-ruby-issue-repro

**Expected behavior**
Event sending not fail when POST request through fetch with non ASCII string.

**Actual behavior**
Event sending fail when POST request through fetch with non ASCII string.

**Environment**
 - Ruby Version: 2.7.3
 - SDK Version:
    - sentry-ruby: 4.5.1
    - sentry-rails: 4.5.1
 - Integration Versions (if any):
    - rails: 5.2.6

**Sentry Config**

This is not necessary but could be helpful.

```ruby
Sentry.init do |config|
  config.send_default_pii = true
  config.debug = true
end
```

**additional info**
I've been searching for a while, and below is my summary for search.

## Why this error occurred?
The direct cause is Rack::Request's body data encoding. It is encoded to ASCII-8BIT, and non ascii string convert to escaped string. And JSON.generate try to convert to UTF-8, but no conversion for escaped string, then raise error.

- If `config.send_default_pii = true` and user send POST request from NON HTML form, sentry-ruby try to send event with request body.
  - https://github.com/getsentry/sentry-ruby/blob/6d2fc60fe21ea3def9da891a470d1a1c6a83760c/sentry-ruby/lib/sentry/interfaces/request.rb#L56-L62
- `Rack::Request#body#read` returns ASCII-8BIT string.
  - NON ASCII-8BIT string converted to an escaped string, such as `\xE5` etc...
- `JSON.generate` (which called at `Sentry::Transport#encode`), try to convert ASCII-8BIT string into UTF-8, then raise this error.

## How to fix this issue
Force-encode request body string to UTF-8 make fix this issue.

For example, in: https://github.com/getsentry/sentry-ruby/blob/6d2fc60fe21ea3def9da891a470d1a1c6a83760c/sentry-ruby/lib/sentry/interfaces/request.rb#L56-L62
```diff
      if request.form_data?
        request.POST
      elsif request.body # JSON requests, etc
-       data = request.body.read(MAX_BODY_LIMIT)
+       data = request.body.read(MAX_BODY_LIMIT).force_encoding(Encoding::UTF_8)
        request.body.rewind
        data
      end
```

But, I'm not sure this is a good solution, because I don't know force_encoding works well in other situations.
- Rails seems to convert (ASCII) URL to UTF-8 at https://github.com/rails/rails/commit/7e504927090362d132d4e315c6f22915050fe5ba, so I believe this solution works well in many cases.

similar: https://github.com/getsentry/sentry-ruby/issues/1303

**Comment by st0012** [comment]
@mkusaka thanks for this detail report! I'll take a look later this week 🙂

**Comment by st0012** [comment]
it's weird that I can reproduce this in your example app (thanks again for this), but not in [`sentry-rails`' example app](https://github.com/getsentry/sentry-ruby/tree/master/sentry-rails/examples/rails-6.0).


```
Processing by DocsController#create as */*
  Parameters: {"title"=>"あ", "content"=>"qweqwe", "doc"=>{"content"=>"qweqwe"}}
Completed 500 Internal Server Error in 1ms (ActiveRecord: 0.0ms)



Event sending failed: "\xE3" from ASCII-8BIT to UTF-8
```

```
Processing by PostsController#create as HTML
  Parameters: {"authenticity_token"=>TOKEN, "post"=>{"title"=>"あ", "content"=>"qweqw"}, "commit"=>"Create Post"}
Completed 500 Internal Server Error in 1ms (ActiveRecord: 0.0ms | Allocations: 1051)


Sending envelope [event] 3b5513966dda4e0c9a618ba955f67e99 to Sentry
```
<img width="80%" alt="截圖 2021-06-25 下午6 58 09" src="https://user-images.githubusercontent.com/5079556/123415062-4bc07d80-d5e7-11eb-9738-5d51a6c03354.png">

I wonder if the difference is the request format?

```
Processing by DocsController#create as */*
Processing by PostsController#create as HTML
```

anyway, I'm working on it now 🙂

**Comment by st0012** [comment]
Fixed with #1484
